### PR TITLE
Make gradient-checkpoint enabling tolerant of models without get_input_embeddings

### PR DESCRIPTION
**Pull request by molbap**
# What does this PR do?

Ad title indicates, #42542 and likely a few other models are broken by merged #41993 . This adds an embedding getter and attempts to test the feature with more coverage.

Basically what it does 

 - Stop hard-failing gradient_checkpointing_enable when a model lacks a get_input_embeddings. We now just call enable_input_require_grads, let it attach hooks where it can, and issue a single warning if no embedding module is found.
 - Simplify enable_input_require_grads (and the InternVL/MLCD and a couple more model overrides/adjustments) by making them responsible for the warning.
-  Adds a big test to make sure all of that works (please take a look)

Should help GC for PEFT adapters for many VLMs hopefully (and normal models too)

### CODE REVIEW

**Comment by molbap** [comment]
added a test as well - but can't find a clean way around the models for which it is _not_ relevant to have a getter method and _not_ causing as many side-effects. WDYT @zucchini-nlp ? kind of stumped (try/excepting at higher level would always work but hides a lot)

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -961,59 +961,78 @@ class EmbeddingAccessMixin:
...
-        else:
-            base_model = getattr(self, "base_model_prefix", None)
-            if base_model is not None:
-                base_model = getattr(self, base_model, None)
-                if base_model is not None and base_model is not self:
-                    return base_model.get_input_embeddings()
+
+        base_model = getattr(self, "base_model_prefix", None)
+        if base_model is not None:
+            base_model = getattr(self, base_model, None)
```
Comment (zucchini-nlp) [inline src/transformers/modeling_utils.py]: nit: `self.base_model` property has the same functionality

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -961,59 +961,78 @@ class EmbeddingAccessMixin:
 
     _input_embed_layer = "embed_tokens"  # default layer that holds input embeddings.
 
-    def get_input_embeddings(self) -> nn.Module:
+    def _get_input_embeddings_no_raise(self) -> Optional[nn.Module]:
```
Comment (zucchini-nlp) [inline src/transformers/modeling_utils.py]: oh interesting, I was assuming the base `get_input_embedding` already returns `None`

**Review by zucchini-nlp — COMMENTED** [review]
> models for which it is not relevant to have a getter method and not causing as many side-effects

Would this also mean that we can't support correctly PEFT and GC with these models, or do they have a custom way to set grad on the inputs? We could raise an error with a better message saying that models doesn't support unless it has a way to get its input embeddings, wdyt?

File: src/transformers/modeling_utils.py (hunk 2 above)
Comment (molbap) [reply src/transformers/modeling_utils.py]: well I ended up in some many little edge cases lol

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (molbap) [reply src/transformers/modeling_utils.py]: true!

**Comment by molbap** [comment]
Yes it's a good idea to raise/inform for downstream users. I reverted a couple things and will update the test so it actually checks that enabling GC works (probably add another test)

File: src/transformers/utils/import_utils.py
Comment (molbap) [inline src/transformers/utils/import_utils.py:1]: This is mostly to fix a broken env situation that can be caused around timm_wrapper (or timm_backbone?) so it protects a few imports

**Comment by molbap** [comment]
I reverted a few models to inner positional embeddings calls as mentioned in https://github.com/huggingface/transformers/pull/38913 . 

Modified a few others models as the test I added (`test_enable_input_require_grads_with_gradient_checkpointing` ) was a bit naive and I was just `continue`-ing, now it's a proper skip if the loss is undefined. 

Hopefully that helps VLMs + GC and does not break adapters

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -1978,14 +1978,18 @@ def make_inputs_require_grads(module, input, output):
...
 
         for module in self.modules():
             if not (isinstance(module, PreTrainedModel) and hasattr(module, "get_input_embeddings")):
                 continue
 
-            input_embeddings = module.get_input_embeddings()
+            try:
+                input_embeddings = module.get_input_embeddings()
+            except NotImplementedError:
+                continue
```
Comment (molbap) [inline src/transformers/modeling_utils.py:2053]: no simple way around this unfortunately

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -1994,11 +1998,18 @@ def make_inputs_require_grads(module, input, output):
...
 
         self._require_grads_hooks = hooks
         if hooks:
             # for BC
             self._require_grads_hook = hooks[0]
+        if not found_embeddings:
+            logger.warning_once(
+                f"{self.__class__.__name__} does not expose input embeddings. Gradients cannot flow back to the token "
+                "embeddings when using adapters or gradient checkpointing. Override `get_input_embeddings` to fully "
+                "support those features."
```
Comment (molbap) [inline src/transformers/modeling_utils.py]: at least we can warn users!

File: src/transformers/models/internvl/modeling_internvl.py [hunk 5]
```diff
@@ -209,10 +209,9 @@ def forward(self, pixel_values: torch.Tensor) -> torch.Tensor:
             )
 
         embeddings = self.projection(pixel_values)
-        patch_height, patch_width = embeddings.shape[2], embeddings.shape[3]
         embeddings = embeddings.flatten(2).transpose(1, 2)
 
-        return embeddings, (patch_height, patch_width)
+        return embeddings
```
Comment (zucchini-nlp) [inline src/transformers/models/internvl/modeling_internvl.py:215]: haha, looks like a left-over from a model shipping and debugging

File: src/transformers/models/altclip/modeling_altclip.py [hunk 6]
```diff
@@ -367,7 +367,6 @@ def feed_forward_chunk(self, attention_output):
         return layer_output
 
 
-# Copied from transformers.models.align.modeling_align.AlignTextEncoder with AlignText->AltRoberta
```
Comment (zucchini-nlp) [inline src/transformers/models/altclip/modeling_altclip.py:370]: why we had to delete a copy?

File: src/transformers/models/mlcd/modeling_mlcd.py [hunk 7]
```diff
@@ -333,9 +334,13 @@ def forward(
         residual = hidden_states
 
         hidden_states = self.layer_norm1(hidden_states)
+        rotary_embeddings = position_embeddings
+        if rotary_position_tensor is not None:
+            rotary_embeddings = (rotary_position_tensor.cos(), rotary_position_tensor.sin())
+
```
Comment (zucchini-nlp) [inline src/transformers/models/mlcd/modeling_mlcd.py]: the `position_embeddings` are already supposed to be present so we don't need the embeddings, isn't it?

File: src/transformers/modeling_utils.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/modeling_utils.py]: nice!

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/modeling_utils.py:2053]: oke, I think with the warning below, it is more explicit

File: src/transformers/modeling_utils.py [hunk 8]
```diff
@@ -2940,7 +2949,10 @@ def gradient_checkpointing_enable(self, gradient_checkpointing_kwargs=None):
                 "Please update to the new format on your modeling file. To use the new format, you need to completely remove the definition of the method `_set_gradient_checkpointing` in your model."
             )
 
-        if getattr(self, "_hf_peft_config_loaded", False):
+        needs_embedding_grads = self.main_input_name == "input_ids"
+        # we use that also to detect whether or not we have to raise if embeddings are missing (the submodel might not have embeddings at all)
+        enable_input_grads = needs_embedding_grads or getattr(self, "_hf_peft_config_loaded", False)
+        if enable_input_grads:
```
Comment (zucchini-nlp) [inline src/transformers/modeling_utils.py:3015]: hmm, for my understanding, why do we always need to enable grads when doing GC training with text models?

File: tests/test_modeling_common.py [hunk 9]
```diff
@@ -906,6 +906,111 @@ def test_peft_gradient_checkpointing_enable_disable(self):
...
+            self.skipTest(reason="ModelTester is not configured to run training tests")
+
+        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
+        if hasattr(config, "use_cache"):
+            config.use_cache = False
+
+        has_verified_model = False
+
+        for model_class in self.all_model_classes:
+            if not getattr(model_class, "supports_gradient_checkpointing", False):
```
Comment (zucchini-nlp) [inline tests/test_modeling_common.py:934]: I see now what you meant earlier, this test has a lot of edge cases

**Review by zucchini-nlp — COMMENTED** [review]
It is a pity that there are so many edge cases to handle. Raising a warning seems like a good solution to not silently skip exceptions

I think there are some unrelated changes with rope, let's revert those before merging

File: src/transformers/models/internvl/modeling_internvl.py (hunk 5 above)
Comment (molbap) [reply src/transformers/models/internvl/modeling_internvl.py:215]: lol yes, not mine, and maybe somebody uses it because it's a public method... but better do that than overriding the grad enabling method

File: src/transformers/models/altclip/modeling_altclip.py (hunk 6 above)
Comment (molbap) [reply src/transformers/models/altclip/modeling_altclip.py:370]: because now both are not exposing their embeddings in exactly the same way, but will see if I can fix it!

File: tests/test_modeling_common.py (hunk 9 above)
Comment (molbap) [reply tests/test_modeling_common.py:934]: yes, it's a bit clunky to have this bool flag but wasn't seeing a simpler option

File: src/transformers/models/internvl/modeling_internvl.py (hunk 5 above)
Comment (zucchini-nlp) [reply src/transformers/models/internvl/modeling_internvl.py:215]: love seeing a fix to a general consistency issue :)

File: src/transformers/models/mlcd/modeling_mlcd.py (hunk 7 above)
Comment (molbap) [reply src/transformers/models/mlcd/modeling_mlcd.py]: that one is a little bit more annoying. I'm trying to revert it but the thing is this model should work except that it's a tuple... and gradient checkpointing does not disable grads on tuples, only on tensors _passed as pos args_ . 

so this was a hacky trick (for the use_reentrant case

File: src/transformers/modeling_utils.py (hunk 8 above)
Comment (molbap) [reply src/transformers/modeling_utils.py:3015]: we don't always, but we do with reentrant checkpointing. IIUC it's not to actualy use these gradients, it's that torch.utils.checkpoint needs at least one input and one output to actually have gradients, else the checkpointed part will not have a gradient.

File: src/transformers/models/mlcd/modeling_mlcd.py (hunk 7 above)
Comment (zucchini-nlp) [reply src/transformers/models/mlcd/modeling_mlcd.py]: weird, we use tuple cos/sin in most LLMs. I'd prefer to skip this model's test instead of fixing by duplicate args, i think it is not used as commonly

**Review by zucchini-nlp — APPROVED** [review]
Thanks again for handling all edge cases, was not an easy one

**Comment by molbap** [comment]
run-slow: align, altclip, chinese_clip, clap, clvp, falcon_mamba, fast_vlm, internvl, layoutlm, layoutlmv3, lilt, mamba, markuplm, mlcd, poolformer, siglip

File: src/transformers/models/mlcd/modeling_mlcd.py (hunk 7 above)
Comment (molbap) [reply src/transformers/models/mlcd/modeling_mlcd.py]: I did skip it yes because it was being annoying :grin: and yes we use them, but they don't usually carry grads, here (in that particular model) they do

File: src/transformers/modeling_utils.py [hunk 10]
```diff
@@ -1994,11 +1996,18 @@ def make_inputs_require_grads(module, input, output):
...
 
         self._require_grads_hooks = hooks
         if hooks:
             # for BC
             self._require_grads_hook = hooks[0]
+        if not found_embeddings:
+            logger.warning_once(
+                f"{self.__class__.__name__} does not expose input embeddings. Gradients cannot flow back to the token "
+                "embeddings when using adapters or gradient checkpointing. Override `get_input_embeddings` to fully "
+                "support those features."
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py]: either that or sometimes just add a `_input_embedding_layer`

File: tests/test_modeling_common.py [hunk 11]
```diff
@@ -906,6 +906,111 @@ def test_peft_gradient_checkpointing_enable_disable(self):
                         m.gradient_checkpointing, f"Module {n} does not have gradient_checkpointing set to False"
                     )
 
+    def test_enable_input_require_grads(self):
+        config, _ = self.model_tester.prepare_config_and_inputs_for_common()
+        for model_class in self.all_model_classes:
+            model = model_class(copy.deepcopy(config))
+            if not hasattr(model, "get_input_embeddings"):
+                continue
```
Comment (ArthurZucker) [inline tests/test_modeling_common.py:915]: why not raise an error instead this way all new models wiil make sure they have this go green before merging?

File: tests/test_modeling_common.py [hunk 12]
```diff
@@ -906,6 +906,111 @@ def test_peft_gradient_checkpointing_enable_disable(self):
...
+            try:
+                model.enable_input_require_grads()
+            except NotImplementedError as error:
+                self.fail(f"enable_input_require_grads raised NotImplementedError for {model_class.__name__}: {error}")
+            finally:
+                model.disable_input_require_grads()
+
+    def test_enable_input_require_grads_with_gradient_checkpointing(self):
+        if not getattr(self.model_tester, "is_training", False):
+            self.skipTest(reason="ModelTester is not configured to run training tests")
```
Comment (ArthurZucker) [inline tests/test_modeling_common.py:925]: If this one is `True` by default for all models sg

**Review by ArthurZucker — APPROVED** [review]
Kudos very nice PR!

File: src/transformers/modeling_utils.py (hunk 10 above)
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py]:
```suggestion
                "support those features, or add the `_input_embedding_layer` attribut with the name of the embedding layer!"
```

File: tests/test_modeling_common.py [hunk 13]
```diff
@@ -906,6 +906,111 @@ def test_peft_gradient_checkpointing_enable_disable(self):
...
+            if loss_tensor is None or not isinstance(loss_tensor, torch.Tensor) or not loss_tensor.requires_grad:
+                model.zero_grad(set_to_none=True)
+                continue
+            loss = loss_tensor.sum()
+            loss.backward()
+
+            grad_after_gc = embedding_param.grad
+            self.assertIsNotNone(
+                grad_after_gc,
+                f"{model_class.__name__} should produce embedding gradients when gradient checkpointing is enabled.",
```
Comment (ArthurZucker) [inline tests/test_modeling_common.py]: if you have an idea of what could cause this to fail, add it!

File: tests/test_modeling_common.py [hunk 14]
```diff
@@ -906,6 +906,111 @@ def test_peft_gradient_checkpointing_enable_disable(self):
...
+                grad_after_gc.abs().sum().item(),
+                0,
+                f"{model_class.__name__} should keep non-zero embedding gradients with gradient checkpointing enabled.",
+            )
+            has_verified_model = True
+
+        if not has_verified_model:
+            self.skipTest(
+                reason="No model with a differentiable loss was available to verify enable_input_require_grads with gradient checkpointing."
+            )
```
Comment (ArthurZucker) [inline tests/test_modeling_common.py:1015]: same for each one of these, let's help the use to fix it!

**Review by ArthurZucker — APPROVED** [review]
First of all thanks! 
This will also fix some TP recompile issues cc @3outeille on `hidden_states=hidden_states`

File: src/transformers/modeling_utils.py (hunk 10 above)
Comment (ArthurZucker) [reply src/transformers/modeling_utils.py]: i mean update the message please

**Review by ArthurZucker — APPROVED** [review]
Very nice! small updates and let's merge

File: tests/test_modeling_common.py (hunk 11 above)
Comment (molbap) [reply tests/test_modeling_common.py:915]: forgot to answer but: this would currently raise for many existing models

File: tests/test_modeling_common.py (hunk 12 above)
Comment (molbap) [reply tests/test_modeling_common.py:925]: yes AFAIK, true for CausalLMTester

**Comment by molbap** [comment]
run-slow: align, altclip, chinese_clip, clap, clvp, falcon_mamba, fast_vlm, internvl, layoutlm, layoutlmv3, lilt, mamba, markuplm, mlcd, poolformer, siglip

**Comment by molbap** [comment]
Comments addressed. Merging and keeping an eye on this :eyes: let's see if something breaks and how
